File: shared/tools/file-ops.ts
/**
 * Unified File Operations Tools for DGMO-DGM Integration
 * Provides read, write, and edit functionality across both systems
 */

import { z } from 'zod';
import * as fs from 'fs/promises';
import * as path from 'path';
import { Tool, ToolContext } from '../types/typescript/tool.types';
import { Result } from '../types/typescript/base.types';

const MAX_READ_SIZE = 250 * 1024;
const DEFAULT_READ_LIMIT = 2000;
const MAX_LINE_LENGTH = 2000;

// ============= READ TOOL =============

export const readInputSchema = z.object({
  filePath: z.string().describe('The path to the file to read'),
  offset: z.number().describe('The line number to start reading from (0-based)').optional(),
  limit: z.number().describe('The number of lines to read (defaults to 2000)').optional(),
});

export type ReadInput = z.infer<typeof readInputSchema>;

export interface ReadOutput {
  content: string;
  lines: number;
  totalLines: number;
  path: string;
}

export async function readFile(
  params: ReadInput,
  context: ToolContext,
): Promise<Result<ReadOutput>> {
  try {
    let filePath = params.filePath;

    // Convert Windows paths to WSL format if needed
    const windowsPathMatch = filePath.match(/^([A-Za-z]):\\(.*)/);
    if (windowsPathMatch) {
      const drive = windowsPathMatch[1].toLowerCase();
      const pathPart = windowsPathMatch[2].replace(/\\/g, '/');
      filePath = `/mnt/${drive}/${pathPart}`;
    }

    if (!path.isAbsolute(filePath)) {
      filePath = path.join(process.cwd(), filePath);
    }

    // Check file exists
    try {
      const stats = await fs.stat(filePath);
      if (stats.size > MAX_READ_SIZE) {
        return {
          success: false,
          error: {
            code: 'FILE_TOO_LARGE',
            message: `File is too large (${stats.size} bytes). Maximum size is ${MAX_READ_SIZE} bytes`,
            recoverable: false,
          },
        };
      }
    } catch (error) {
      return {
        success: false,
        error: {
          code: 'FILE_NOT_FOUND',
          message: `File not found: ${filePath}`,
          recoverable: false,
        },
      };
    }

    const content = await fs.readFile(filePath, 'utf-8');
    const lines = content.split('\n');
    const limit = params.limit ?? DEFAULT_READ_LIMIT;
    const offset = params.offset || 0;

    const selectedLines = lines.slice(offset, offset + limit);
    const formattedLines = selectedLines.map((line, index) => {
      const lineNum = (index + offset + 1).toString().padStart(5, '0');
      const truncatedLine =
        line.length > MAX_LINE_LENGTH ? line.substring(0, MAX_LINE_LENGTH) + '...' : line;
      return `${lineNum}| ${truncatedLine}`;
    });

    const output: ReadOutput = {
      content: formattedLines.join('\n'),
      lines: selectedLines.length,
      totalLines: lines.length,
      path: filePath,
    };

    return {
      success: true,
      data: output,
      metadata: {
        id: context.messageId,
        version: '1.0.0',
        timestamp: new Date().toISOString(),
        source: 'unified-read',
      },
    };
  } catch (error) {
    return {
      success: false,
      error: {
        code: 'READ_ERROR',
        message: error instanceof Error ? error.message : String(error),
        recoverable: false,
        details: error,
      },
    };
  }
}

// ============= WRITE TOOL =============

export const writeInputSchema = z.object({
  filePath: z
    .string()
    .describe('The absolute path to the file to write (must be absolute, not relative)'),
  content: z.string().describe('The content to write to the file'),
});

export type WriteInput = z.infer<typeof writeInputSchema>;

export interface WriteOutput {
  path: string;
  exists: boolean;
  size: number;
}

export async function writeFile(
  params: WriteInput,
  context: ToolContext,
): Promise<Result<WriteOutput>> {
  try {
    let filePath = params.filePath;

    if (!path.isAbsolute(filePath)) {
      filePath = path.join(process.cwd(), filePath);
    }

    // Check if file exists
    let exists = false;
    try {
      await fs.stat(filePath);
      exists = true;
    } catch {
      // File doesn't exist
    }

    // Ensure directory exists
    const dir = path.dirname(filePath);
    await fs.mkdir(dir, { recursive: true });

    // Write file
    await fs.writeFile(filePath, params.content, 'utf-8');

    const stats = await fs.stat(filePath);

    const output: WriteOutput = {
      path: filePath,
      exists,
      size: stats.size,
    };

    return {
      success: true,
      data: output,
      metadata: {
        id: context.messageId,
        version: '1.0.0',
        timestamp: new Date().toISOString(),
        source: 'unified-write',
      },
    };
  } catch (error) {
    return {
      success: false,
      error: {
        code: 'WRITE_ERROR',
        message: error instanceof Error ? error.message : String(error),
        recoverable: false,
        details: error,
      },
    };
  }
}

// ============= EDIT TOOL =============

export const editInputSchema = z.object({
  filePath: z.string().describe('The absolute path to the file to modify'),
  oldString: z.string().describe('The text to replace'),
  newString: z.string().describe('The text to replace it with (must be different from old_string)'),
  replaceAll: z
    .boolean()
    .optional()
    .describe('Replace all occurrences of old_string (default false)'),
});

export type EditInput = z.infer<typeof editInputSchema>;

export interface EditOutput {
  path: string;
  replacements: number;
  diff: string;
}

export async function editFile(
  params: EditInput,
  context: ToolContext,
): Promise<Result<EditOutput>> {
  try {
    if (params.oldString === params.newString) {
      return {
        success: false,
        error: {
          code: 'INVALID_EDIT',
          message: 'oldString and newString must be different',
          recoverable: false,
        },
      };
    }

    let filePath = params.filePath;

    if (!path.isAbsolute(filePath)) {
      filePath = path.join(process.cwd(), filePath);
    }

    // Read file
    let content: string;
    try {
      content = await fs.readFile(filePath, 'utf-8');
    } catch (error) {
      return {
        success: false,
        error: {
          code: 'FILE_NOT_FOUND',
          message: `File not found: ${filePath}`,
          recoverable: false,
        },
      };
    }

    // Perform replacement
    let newContent: string;
    let replacements = 0;

    if (params.replaceAll) {
      const regex = new RegExp(escapeRegExp(params.oldString), 'g');
      newContent = content.replace(regex, () => {
        replacements++;
        return params.newString;
      });
    } else {
      const index = content.indexOf(params.oldString);
      if (index !== -1) {
        newContent =
          content.substring(0, index) +
          params.newString +
          content.substring(index + params.oldString.length);
        replacements = 1;
      } else {
        newContent = content;
      }
    }

    if (replacements === 0) {
      return {
        success: false,
        error: {
          code: 'STRING_NOT_FOUND',
          message: `String not found in file: ${params.oldString}`,
          recoverable: false,
        },
      };
    }

    // Write file
    await fs.writeFile(filePath, newContent, 'utf-8');

    // Create simple diff
    const diff = createSimpleDiff(content, newContent, filePath);

    const output: EditOutput = {
      path: filePath,
      replacements,
      diff,
    };

    return {
      success: true,
      data: output,
      metadata: {
        id: context.messageId,
        version: '1.0.0',
        timestamp: new Date().toISOString(),
        source: 'unified-edit',
      },
    };
  } catch (error) {
    return {
      success: false,
      error: {
        code: 'EDIT_ERROR',
        message: error instanceof Error ? error.message : String(error),
        recoverable: false,
        details: error,
      },
    };
  }
}

// Helper functions
function escapeRegExp(string: string): string {
  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function createSimpleDiff(oldContent: string, newContent: string, filePath: string): string {
  const oldLines = oldContent.split('\n');
  const newLines = newContent.split('\n');

  let diff = `--- ${filePath}\n+++ ${filePath}\n`;

  // Simple line-by-line comparison
  const maxLines = Math.max(oldLines.length, newLines.length);
  for (let i = 0; i < maxLines; i++) {
    if (i >= oldLines.length) {
      diff += `+${newLines[i]}\n`;
    } else if (i >= newLines.length) {
      diff += `-${oldLines[i]}\n`;
    } else if (oldLines[i] !== newLines[i]) {
      diff += `-${oldLines[i]}\n`;
      diff += `+${newLines[i]}\n`;
    }
  }

  return diff;
}

// Tool definitions
export const readTool: Tool = {
  id: 'read',
  name: 'read',
  description: 'Read contents of a file with line numbers',
  version: '1.0.0',
  category: 'file-system',
  language: 'typescript',
  inputSchema: {
    type: 'object',
    properties: {
      filePath: { type: 'string', description: 'The path to the file to read' },
      offset: { type: 'number', description: 'The line number to start reading from (0-based)' },
      limit: { type: 'number', description: 'The number of lines to read (defaults to 2000)' },
    },
    required: ['filePath'],
  },
  outputSchema: {
    type: 'object',
    properties: {
      content: { type: 'string' },
      lines: { type: 'number' },
      totalLines: { type: 'number' },
      path: { type: 'string' },
    },
    required: ['content', 'lines', 'totalLines', 'path'],
  },
  metadata: {
    id: 'unified-read',
    version: '1.0.0',
    timestamp: new Date().toISOString(),
    source: 'shared',
  },
};

export const writeTool: Tool = {
  id: 'write',
  name: 'write',
  description: 'Write content to a file',
  version: '1.0.0',
  category: 'file-system',
  language: 'typescript',
  inputSchema: {
    type: 'object',
    properties: {
      filePath: { type: 'string', description: 'The absolute path to the file to write' },
      content: { type: 'string', description: 'The content to write to the file' },
    },
    required: ['filePath', 'content'],
  },
  outputSchema: {
    type: 'object',
    properties: {
      path: { type: 'string' },
      exists: { type: 'boolean' },
      size: { type: 'number' },
    },
    required: ['path', 'exists', 'size'],
  },
  metadata: {
    id: 'unified-write',
    version: '1.0.0',
    timestamp: new Date().toISOString(),
    source: 'shared',
  },
};

export const editTool: Tool = {
  id: 'edit',
  name: 'edit',
  description: 'Edit a file by replacing text',
  version: '1.0.0',
  category: 'file-system',
  language: 'typescript',
  inputSchema: {
    type: 'object',
    properties: {
      filePath: { type: 'string', description: 'The absolute path to the file to modify' },
      oldString: { type: 'string', description: 'The text to replace' },
      newString: { type: 'string', description: 'The text to replace it with' },
      replaceAll: { type: 'boolean', description: 'Replace all occurrences (default false)' },
    },
    required: ['filePath', 'oldString', 'newString'],
  },
  outputSchema: {
    type: 'object',
    properties: {
      path: { type: 'string' },
      replacements: { type: 'number' },
      diff: { type: 'string' },
    },
    required: ['path', 'replacements', 'diff'],
  },
  metadata: {
    id: 'unified-edit',
    version: '1.0.0',
    timestamp: new Date().toISOString(),
    source: 'shared',
  },
};

// Tool handlers
export async function readToolHandler(
  input: ReadInput,
  context: ToolContext,
): Promise<Result<any>> {
  const result = await readFile(input, context);

  if (result.success && result.data) {
    let output = '<file>\n';
    output += result.data.content;

    if (result.data.totalLines > result.data.lines + (input.offset || 0)) {
      output += `\n\n(File has more lines. Use 'offset' parameter to read beyond line ${
        result.data.lines + (input.offset || 0)
      })`;
    }
    output += '\n</file>';

    return {
      ...result,
      data: {
        output,
        metadata: {
          preview: result.data.content.split('\n').slice(0, 20).join('\n'),
          title: path.basename(result.data.path),
        },
      },
    };
  }

  return result;
}

export async function writeToolHandler(
  input: WriteInput,
  context: ToolContext,
): Promise<Result<any>> {
  const result = await writeFile(input, context);

  if (result.success && result.data) {
    return {
      ...result,
      data: {
        output: '',
        metadata: {
          filepath: result.data.path,
          exists: result.data.exists,
          title: path.basename(result.data.path),
        },
      },
    };
  }

  return result;
}

export async function editToolHandler(
  input: EditInput,
  context: ToolContext,
): Promise<Result<any>> {
  const result = await editFile(input, context);

  if (result.success && result.data) {
    return {
      ...result,
      data: {
        output: result.data.diff,
        metadata: {
          filepath: result.data.path,
          replacements: result.data.replacements,
          title: path.basename(result.data.path),
        },
      },
    };
  }

  return result;
}

// DGMO compatibility exports
export const ReadTool = {
  id: 'read',
  description: readTool.description,
  parameters: readInputSchema,
  async execute(params: ReadInput, ctx: any) {
    const context: ToolContext = {
      sessionId: ctx.sessionID,
      messageId: ctx.messageID,
      abortSignal: ctx.abort,
      timeout: 30000,
      metadata: new Map(),
      environment: process.env as Record<string, string>,
      logger: {
        debug: (message: string, data?: any) => console.debug(message, data),
        info: (message: string, data?: any) => console.info(message, data),
        warn: (message: string, data?: any) => console.warn(message, data),
        error: (message: string, error?: any) => console.error(message, error),
        metric: (name: string, value: number, tags?: Record<string, string>) =>
          console.log(`METRIC: ${name}=${value}`, tags),
      },
    };

    const result = await readToolHandler(params, context);

    if (result.success) {
      return result.data;
    }

    throw new Error(result.error?.message || 'Read failed');
  },
};

export const WriteTool = {
  id: 'write',
  description: writeTool.description,
  parameters: writeInputSchema,
  async execute(params: WriteInput, ctx: any) {
    const context: ToolContext = {
      sessionId: ctx.sessionID,
      messageId: ctx.messageID,
      abortSignal: ctx.abort,
      timeout: 30000,
      metadata: new Map(),
      environment: process.env as Record<string, string>,
      logger: {
        debug: (message: string, data?: any) => console.debug(message, data),
        info: (message: string, data?: any) => console.info(message, data),
        warn: (message: string, data?: any) => console.warn(message, data),
        error: (message: string, error?: any) => console.error(message, error),
        metric: (name: string, value: number, tags?: Record<string, string>) =>
          console.log(`METRIC: ${name}=${value}`, tags),
      },
    };

    const result = await writeToolHandler(params, context);

    if (result.success) {
      return result.data;
    }

    throw new Error(result.error?.message || 'Write failed');
  },
};

export const EditTool = {
  id: 'edit',
  description: editTool.description,
  parameters: editInputSchema,
  async execute(params: EditInput, ctx: any) {
    const context: ToolContext = {
      sessionId: ctx.sessionID,
      messageId: ctx.messageID,
      abortSignal: ctx.abort,
      timeout: 30000,
      metadata: new Map(),
      environment: process.env as Record<string, string>,
      logger: {
        debug: (message: string, data?: any) => console.debug(message, data),
        info: (message: string, data?: any) => console.info(message, data),
        warn: (message: string, data?: any) => console.warn(message, data),
        error: (message: string, error?: any) => console.error(message, error),
        metric: (name: string, value: number, tags?: Record<string, string>) =>
          console.log(`METRIC: ${name}=${value}`, tags),
      },
    };

    const result = await editToolHandler(params, context);

    if (result.success) {
      return result.data;
    }

    throw new Error(result.error?.message || 'Edit failed');
  },
};

File: shared/tools/file_ops.py
"""
Unified File Operations Tools for DGMO-DGM Integration
Provides read, write, and edit functionality across both systems
"""

import os
import re
from pathlib import Path
from typing import Dict, Any, Optional
from datetime import datetime

from ..types.python.tool import Tool, ToolContext, ToolCategory, Language
from ..types.python.base import Result, ErrorInfo

MAX_READ_SIZE = 250 * 1024
DEFAULT_READ_LIMIT = 2000
MAX_LINE_LENGTH = 2000


# ============= READ TOOL =============

async def read_file(params: Dict[str, Any], context: ToolContext) -> Result:
    """Read file with line numbers"""
    try:
        file_path = params['filePath']
        
        # Convert Windows paths to WSL format if needed
        windows_match = re.match(r'^([A-Za-z]):\\(.*)$', file_path)
        if windows_match:
            drive = windows_match.group(1).lower()
            path_part = windows_match.group(2).replace('\\', '/')
            file_path = f'/mnt/{drive}/{path_part}'
        
        if not os.path.isabs(file_path):
            file_path = os.path.abspath(file_path)
        
        # Check file exists
        if not os.path.exists(file_path):
            return Result(
                success=False,
                error=ErrorInfo(
                    code='FILE_NOT_FOUND',
                    message=f'File not found: {file_path}',
                    recoverable=False
                )
            )
        
        # Check file size
        file_size = os.path.getsize(file_path)
        if file_size > MAX_READ_SIZE:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='FILE_TOO_LARGE',
                    message=f'File is too large ({file_size} bytes). Maximum size is {MAX_READ_SIZE} bytes',
                    recoverable=False
                )
            )
        
        # Read file
        with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
            content = f.read()
        
        lines = content.split('\n')
        limit = params.get('limit', DEFAULT_READ_LIMIT)
        offset = params.get('offset', 0)
        
        selected_lines = lines[offset:offset + limit]
        formatted_lines = []
        
        for i, line in enumerate(selected_lines):
            line_num = str(i + offset + 1).zfill(5)
            truncated_line = line[:MAX_LINE_LENGTH] + '...' if len(line) > MAX_LINE_LENGTH else line
            formatted_lines.append(f'{line_num}| {truncated_line}')
        
        output = {
            'content': '\n'.join(formatted_lines),
            'lines': len(selected_lines),
            'totalLines': len(lines),
            'path': file_path
        }
        
        return Result(
            success=True,
            data=output,
            metadata={
                'id': context.message_id,
                'version': '1.0.0',
                'timestamp': datetime.now().isoformat(),
                'source': 'unified-read'
            }
        )
        
    except Exception as e:
        return Result(
            success=False,
            error=ErrorInfo(
                code='READ_ERROR',
                message=str(e),
                recoverable=False,
                details={'error': str(e)}
            )
        )


# ============= WRITE TOOL =============

async def write_file(params: Dict[str, Any], context: ToolContext) -> Result:
    """Write content to a file"""
    try:
        file_path = params['filePath']
        content = params['content']
        
        if not os.path.isabs(file_path):
            file_path = os.path.abspath(file_path)
        
        # Check if file exists
        exists = os.path.exists(file_path)
        
        # Ensure directory exists
        directory = os.path.dirname(file_path)
        os.makedirs(directory, exist_ok=True)
        
        # Write file
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        
        file_size = os.path.getsize(file_path)
        
        output = {
            'path': file_path,
            'exists': exists,
            'size': file_size
        }
        
        return Result(
            success=True,
            data=output,
            metadata={
                'id': context.message_id,
                'version': '1.0.0',
                'timestamp': datetime.now().isoformat(),
                'source': 'unified-write'
            }
        )
        
    except Exception as e:
        return Result(
            success=False,
            error=ErrorInfo(
                code='WRITE_ERROR',
                message=str(e),
                recoverable=False,
                details={'error': str(e)}
            )
        )


# ============= EDIT TOOL =============

async def edit_file(params: Dict[str, Any], context: ToolContext) -> Result:
    """Edit a file by replacing text"""
    try:
        file_path = params['filePath']
        old_string = params['oldString']
        new_string = params['newString']
        replace_all = params.get('replaceAll', False)
        
        if old_string == new_string:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='INVALID_EDIT',
                    message='oldString and newString must be different',
                    recoverable=False
                )
            )
        
        if not os.path.isabs(file_path):
            file_path = os.path.abspath(file_path)
        
        # Read file
        if not os.path.exists(file_path):
            return Result(
                success=False,
                error=ErrorInfo(
                    code='FILE_NOT_FOUND',
                    message=f'File not found: {file_path}',
                    recoverable=False
                )
            )
        
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Perform replacement
        replacements = 0
        if replace_all:
            new_content = content
            while old_string in new_content:
                new_content = new_content.replace(old_string, new_string, 1)
                replacements += 1
        else:
            if old_string in content:
                new_content = content.replace(old_string, new_string, 1)
                replacements = 1
            else:
                new_content = content
        
        if replacements == 0:
            return Result(
                success=False,
                error=ErrorInfo(
                    code='STRING_NOT_FOUND',
                    message=f'String not found in file: {old_string}',
                    recoverable=False
                )
            )
        
        # Write file
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        
        # Create simple diff
        diff = create_simple_diff(content, new_content, file_path)
        
        output = {
            'path': file_path,
            'replacements': replacements,
            'diff': diff
        }
        
        return Result(
            success=True,
            data=output,
            metadata={
                'id': context.message_id,
                'version': '1.0.0',
                'timestamp': datetime.now().isoformat(),
                'source': 'unified-edit'
            }
        )
        
    except Exception as e:
        return Result(
            success=False,
            error=ErrorInfo(
                code='EDIT_ERROR',
                message=str(e),
                recoverable=False,
                details={'error': str(e)}
            )
        )


def create_simple_diff(old_content: str, new_content: str, file_path: str) -> str:
    """Create a simple diff between old and new content"""
    old_lines = old_content.split('\n')
    new_lines = new_content.split('\n')
    
    diff = f'--- {file_path}\n+++ {file_path}\n'
    
    # Simple line-by-line comparison
    max_lines = max(len(old_lines), len(new_lines))
    for i in range(max_lines):
        if i >= len(old_lines):
            diff += f'+{new_lines[i]}\n'
        elif i >= len(new_lines):
            diff += f'-{old_lines[i]}\n'
        elif old_lines[i] != new_lines[i]:
            diff += f'-{old_lines[i]}\n'
            diff += f'+{new_lines[i]}\n'
    
    return diff


# Tool definitions
read_tool = Tool(
    id='read',
    name='read',
    description='Read contents of a file with line numbers',
    version='1.0.0',
    category=ToolCategory.FILE_SYSTEM,
    language=Language.PYTHON,
    input_schema={
        'type': 'object',
        'properties': {
            'filePath': {'type': 'string', 'description': 'The path to the file to read'},
            'offset': {'type': 'number', 'description': 'The line number to start reading from (0-based)'},
            'limit': {'type': 'number', 'description': 'The number of lines to read (defaults to 2000)'}
        },
        'required': ['filePath']
    },
    output_schema={
        'type': 'object',
        'properties': {
            'content': {'type': 'string'},
            'lines': {'type': 'number'},
            'totalLines': {'type': 'number'},
            'path': {'type': 'string'}
        },
        'required': ['content', 'lines', 'totalLines', 'path']
    },
    metadata={
        'id': 'unified-read',
        'version': '1.0.0',
        'timestamp': datetime.now().isoformat(),
        'source': 'shared'
    }
)

write_tool = Tool(
    id='write',
    name='write',
    description='Write content to a file',
    version='1.0.0',
    category=ToolCategory.FILE_SYSTEM,
    language=Language.PYTHON,
    input_schema={
        'type': 'object',
        'properties': {
            'filePath': {'type': 'string', 'description': 'The absolute path to the file to write'},
            'content': {'type': 'string', 'description': 'The content to write to the file'}
        },
        'required': ['filePath', 'content']
    },
    output_schema={
        'type': 'object',
        'properties': {
            'path': {'type': 'string'},
            'exists': {'type': 'boolean'},
            'size': {'type': 'number'}
        },
        'required': ['path', 'exists', 'size']
    },
    metadata={
        'id': 'unified-write',
        'version': '1.0.0',
        'timestamp': datetime.now().isoformat(),
        'source': 'shared'
    }
)

edit_tool = Tool(
    id='edit',
    name='edit',
    description='Edit a file by replacing text',
    version='1.0.0',
    category=ToolCategory.FILE_SYSTEM,
    language=Language.PYTHON,
    input_schema={
        'type': 'object',
        'properties': {
            'filePath': {'type': 'string', 'description': 'The absolute path to the file to modify'},
            'oldString': {'type': 'string', 'description': 'The text to replace'},
            'newString': {'type': 'string', 'description': 'The text to replace it with'},
            'replaceAll': {'type': 'boolean', 'description': 'Replace all occurrences (default false)'}
        },
        'required': ['filePath', 'oldString', 'newString']
    },
    output_schema={
        'type': 'object',
        'properties': {
            'path': {'type': 'string'},
            'replacements': {'type': 'number'},
            'diff': {'type': 'string'}
        },
        'required': ['path', 'replacements', 'diff']
    },
    metadata={
        'id': 'unified-edit',
        'version': '1.0.0',
        'timestamp': datetime.now().isoformat(),
        'source': 'shared'
    }
)


# Tool handlers
async def read_tool_handler(input_data: Dict[str, Any], context: ToolContext) -> Result:
    """Read tool handler for unified registry"""
    result = await read_file(input_data, context)
    
    if result.success and result.data:
        output = '<file>\n'
        output += result.data['content']
        
        if result.data['totalLines'] > result.data['lines'] + input_data.get('offset', 0):
            output += f"\n\n(File has more lines. Use 'offset' parameter to read beyond line {
                result.data['lines'] + input_data.get('offset', 0)
            })"
        output += '\n</file>'
        
        return Result(
            success=True,
            data={
                'output': output,
                'metadata': {
                    'preview': '\n'.join(result.data['content'].split('\n')[:20]),
                    'title': os.path.basename(result.data['path'])
                }
            },
            metadata=result.metadata
        )
    
    return result


async def write_tool_handler(input_data: Dict[str, Any], context: ToolContext) -> Result:
    """Write tool handler for unified registry"""
    result = await write_file(input_data, context)
    
    if result.success and result.data:
        return Result(
            success=True,
            data={
                'output': '',
                'metadata': {
                    'filepath': result.data['path'],
                    'exists': result.data['exists'],
                    'title': os.path.basename(result.data['path'])
                }
            },
            metadata=result.metadata
        )
    
    return result


async def edit_tool_handler(input_data: Dict[str, Any], context: ToolContext) -> Result:
    """Edit tool handler for unified registry"""
    result = await edit_file(input_data, context)
    
    if result.success and result.data:
        return Result(
            success=True,
            data={
                'output': result.data['diff'],
                'metadata': {
                    'filepath': result.data['path'],
                    'replacements': result.data['replacements'],
                    'title': os.path.basename(result.data['path'])
                }
            },
            metadata=result.metadata
        )
    
    return result


# DGM compatibility functions
async def tool_function_async_read(params: Dict[str, Any], context: Optional[ToolContext] = None) -> Dict[str, Any]:
    """DGM-compatible read tool function"""
    if not context:
        context = ToolContext(
            session_id='dgm-session',
            message_id='dgm-message',
            timeout=30,
            metadata={}
        )
    
    result = await read_tool_handler(params, context)
    
    if result.success:
        return result.data
    else:
        raise Exception(result.error.message if result.error else 'Read failed')


async def tool_function_async_write(params: Dict[str, Any], context: Optional[ToolContext] = None) -> Dict[str, Any]:
    """DGM-compatible write tool function"""
    if not context:
        context = ToolContext(
            session_id='dgm-session',
            message_id='dgm-message',
            timeout=30,
            metadata={}
        )
    
    result = await write_tool_handler(params, context)
    
    if result.success:
        return result.data
    else:
        raise Exception(result.error.message if result.error else 'Write failed')


async def tool_function_async_edit(params: Dict[str, Any], context: Optional[ToolContext] = None) -> Dict[str, Any]:
    """DGM-compatible edit tool function"""
    if not context:
        context = ToolContext(
            session_id='dgm-session',
            message_id='dgm-message',
            timeout=30,
            metadata={}
        )
    
    result = await edit_tool_handler(params, context)
    
    if result.success:
        return result.data
    else:
        raise Exception(result.error.message if result.error else 'Edit failed')


def tool_info_read() -> Dict[str, Any]:
    """Return read tool information for DGM registration"""
    return {
        'name': 'read',
        'description': read_tool.description,
        'input_schema': read_tool.input_schema,
        'output_schema': read_tool.output_schema,
        'category': read_tool.category.value,
        'tags': ['file', 'read', 'content']
    }


def tool_info_write() -> Dict[str, Any]:
    """Return write tool information for DGM registration"""
    return {
        'name': 'write',
        'description': write_tool.description,
        'input_schema': write_tool.input_schema,
        'output_schema': write_tool.output_schema,
        'category': write_tool.category.value,
        'tags': ['file', 'write', 'create']
    }


def tool_info_edit() -> Dict[str, Any]:
    """Return edit tool information for DGM registration"""
    return {
        'name': 'edit',
        'description': edit_tool.description,
        'input_schema': edit_tool.input_schema,
        'output_schema': edit_tool.output_schema,
        'category': edit_tool.category.value,
        'tags': ['file', 'edit', 'replace']
    }